http_service/app.py

```diff
@@ -3,37 +3,80 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/.
 
+import json
 import logging
 import os
+from collections import UserDict, defaultdict
+from multiprocessing.pool import AsyncResult, Pool
 
-from flask import Flask, current_app, jsonify, request
+from flask import Flask, jsonify, request
 
-from bugbug import bugzilla
-
-from .models import load_model
+from .models import classify_bug
 
 API_TOKEN = "X-Api-Key"
 
 application = Flask(__name__)
 
-bugzilla.set_token(os.environ.get("BUGBUG_BUGZILLA_TOKEN"))
+BUGZILLA_TOKEN = os.environ.get("BUGBUG_BUGZILLA_TOKEN")
 
 logging.basicConfig(level=logging.INFO)
 LOGGER = logging.getLogger()
 
 
-def get_model(model):
-    attribute = f"bugbug_model_{model}"
+POOL = Pool(4)
+
+
+class Database(UserDict):
+    # We might want to customize the database with cache invalidation or plug
+    # it into a real database / caching service
+    pass
+
+
+MODELS_DB = defaultdict(Database)
+
+
+def schedule_bug_classification(model_name, bug_id):
+    """ Schedule the classification of a single bug_id
+    """
+    res = POOL.apply_async(classify_bug, (model_name, bug_id, BUGZILLA_TOKEN))
```

Comment:
For performance reasons, it'd be better to do the classification by batches, but this would increase the complexity of the architecture.
We can do it as a follow-up.

Reply:
I tried to do the classifications for all given bugs but the return structure of `classify_bug` is not very row-oriented friendly. How much performance gain could we expect doing batches?

Reply:
I'm not sure, maybe we can try taking a model and 100 bugs and test how long it takes in batches of 1, 5, 10, 100.

Reply:
I checked quickly what the overhead was and it was non-negligeable, it took 9s to process the 100 bugs in one batch (with a preloaded model) and 29s to process them in batchs of size 10. I tried with a batch of size 1 but the benchmark was still running after 22 minutes.

I checked the RQ Worker implementation and the worker is forking right before handling a job and is designed to handle only one job. That means that we cannot cache the model in memory at least without some customization of the Worker. It should also be possible to have a custom Worker able to process several RQ tasks at once instead of one by one, so we could still schedule them one by one, but I'm not sure how much work is it needed.

@marco-c Do you think we will have lot of requests in the first time? It takes about 2 seconds to classify a bug right now with the model loading.

Reply:
I tried with the scripts we are running in relman-auto-nag. The first needed to classify 443 bugs, the second 453 bugs, the third 1380 bugs.
Usually the numbers will be smaller though (e.g. the third is run every hour, but it only needs to classify bugs that it hasn't classified yet, so the first run will classify 1380 bugs, the second run only bugs that weren't classified before, so < 10).